### PR TITLE
Stack depth finder

### PR DESCRIPTION
**Pull request by TyOverby**
```
C:\Users\tyoverby\workspace\roslyn>csi .\src\Test\Perf\StackDepthTest\runner.csx .\Binaries\Debug\Exes\StackDepthTest\StackDepthTest.exe 3000
Running 1500: fail
Running 750: fail
Running 375: fail
Running 187: pass
Running 281: pass
Running 328: fail
Running 304: fail
Running 292: pass
Running 298: fail
Running 295: fail
Running 293: pass
Running 294: pass
Break even point: 294
```

### CODE REVIEW

**Comment by TyOverby** [comment]
@jaredpar @jcouv

**Comment by TyOverby** [comment]
Interestingly, this stack depth finder finds lower stack depth than we've shown in the tests.

**Comment by jaredpar** [comment]
> Interestingly, this stack depth finder finds lower stack depth than we've shown in the tests.

I would actually expect that.  In the test we create a thread with a known (suggestion) stack size and execute directly on it.  Hence the best possible scenario.

In the actual compiler our binder gets executed in a thread pool and hence loses a bit of the stack from the start.

File: src/Test/Perf/StackDepthTest/Program.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+
```
Comment (jaredpar) [inline src/Test/Perf/StackDepthTest/Program.cs]: Remove unnecessayr blank lines.

File: src/Test/Perf/StackDepthTest/Program.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
+﻿using Microsoft.CodeAnalysis;
```
Comment (jaredpar) [inline src/Test/Perf/StackDepthTest/Program.cs]: Add copyright headers to new files.

File: Roslyn.sln [hunk 3]
```diff
@@ -368,6 +368,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "TestUtilities.CoreClr", "sr
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "VisualStudioSetup.Dependencies", "src\VisualStudio\Setup.Dependencies\VisualStudioSetup.Dependencies.csproj", "{1688E1E5-D510-4E06-86F3-F8DB10B1393D}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "StackDepthTest", "src\test\Perf\StackDepthTest\StackDepthTest.csproj", "{F040CEC5-5E11-4DBD-9F6A-250478E28177}"
```
Comment (jaredpar) [inline Roslyn.sln]: The lower case `test` here makes me nervous.  Can you verify that the git tree for the new files actually uses Test vs. test.  Very laborous to undo afterwards.

File: src/Test/Perf/StackDepthTest/App.config [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+        <assemblyIdentity name="System.IO.FileSystem" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-4.0.2.0" newVersion="4.0.2.0" />
+      </dependentAssembly>
+      <dependentAssembly>
+        <assemblyIdentity name="System.Collections.Immutable" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-1.2.1.0" newVersion="1.2.1.0" />
+      </dependentAssembly>
+    </assemblyBinding>
+  </runtime>
+</configuration>
```
Comment (jaredpar) [inline src/Test/Perf/StackDepthTest/App.config]: Can you try auto-generating redirects here vs. checking in the config file?

File: src/Test/Perf/StackDepthTest/Program.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        public enum ErrorModes : uint
+        {
+            SYSTEM_DEFAULT = 0x0,
+            SEM_FAILCRITICALERRORS = 0x0001,
+            SEM_NOALIGNMENTFAULTEXCEPT = 0x0004,
+            SEM_NOGPFAULTERRORBOX = 0x0002,
+            SEM_NOOPENFILEERRORBOX = 0x8000
+        }
+
+        static int Main(string[] args)
```
Comment (jaredpar) [inline src/Test/Perf/StackDepthTest/Program.cs]: Need a accessibility modifier.

File: src/Test/Perf/StackDepthTest/Program.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+using Microsoft.CodeAnalysis.CSharp;
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+
+
+namespace OverflowSensitivity
+{
+    class Program
```
Comment (jaredpar) [inline src/Test/Perf/StackDepthTest/Program.cs]: Need a accessibility modifier.

File: src/Test/Perf/StackDepthTest/Program.cs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        [DllImport("kernel32.dll")]
+        static extern ErrorModes SetErrorMode(ErrorModes uMode);
```
Comment (jaredpar) [inline src/Test/Perf/StackDepthTest/Program.cs]: Need a accessibility modifier.

File: src/Test/Perf/StackDepthTest/Program.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                    return 0;
+            }
+            else
+            {
+                Console.WriteLine($"Could not parse {args[0]}");
+                return -1;
+            }
+    }
+
+    private static string MakeCode(int depth)
```
Comment (jaredpar) [inline src/Test/Perf/StackDepthTest/Program.cs]: Looks like your indentiation is off here.

File: src/Test/Perf/StackDepthTest/Program.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                Console.WriteLine("You must pass an integer argument in to this program.");
+                return -1;
+            }
+
+            Console.WriteLine($"Running in {IntPtr.Size * 8}-bit mode");
+
+            if(int.TryParse(args[0], out var i))
+            {
+                    CompileCode(MakeCode(i));
```
Comment (jaredpar) [inline src/Test/Perf/StackDepthTest/Program.cs]: Lots of indent issues in this file.

File: src/Test/Perf/StackDepthTest/Program.cs (hunk 2 above)
Comment (TyOverby) [inline src/Test/Perf/StackDepthTest/Program.cs]: Done

File: src/Test/Perf/StackDepthTest/Program.cs (hunk 1 above)
Comment (TyOverby) [inline src/Test/Perf/StackDepthTest/Program.cs]: Done

File: src/Test/Perf/StackDepthTest/Program.cs (hunk 6 above)
Comment (TyOverby) [inline src/Test/Perf/StackDepthTest/Program.cs]: Done

File: src/Test/Perf/StackDepthTest/Program.cs (hunk 7 above)
Comment (TyOverby) [inline src/Test/Perf/StackDepthTest/Program.cs]: Done

File: src/Test/Perf/StackDepthTest/Program.cs (hunk 5 above)
Comment (TyOverby) [inline src/Test/Perf/StackDepthTest/Program.cs]: Done

File: src/Test/Perf/StackDepthTest/Program.cs (hunk 9 above)
Comment (TyOverby) [inline src/Test/Perf/StackDepthTest/Program.cs]: I hit auto-format and it changed some things.  Looks ok now.

File: src/Test/Perf/StackDepthTest/Program.cs (hunk 8 above)
Comment (TyOverby) [inline src/Test/Perf/StackDepthTest/Program.cs]: Same as above

File: Roslyn.sln (hunk 3 above)
Comment (TyOverby) [inline Roslyn.sln]: Hmm, it looks like it looks like it should be "Test".

File: src/Test/Perf/StackDepthTest/App.config (hunk 4 above)
Comment (TyOverby) [inline src/Test/Perf/StackDepthTest/App.config]: Oh wow, I didn't even know that existed.  That's super handy.

**Comment by TyOverby** [comment]
@jaredpar @tannergooding 

I'm guessing these warnings failed the builds?  Know how to fix them?

```
15:04:57 WARNING: Dependency specified was System.Collections.Immutable (>= 1.2.1) but ended up with System.Collections.Immutable 1.3.0.
15:04:57 WARNING: Dependency specified was System.Collections.Immutable (>= 1.2.1) but ended up with System.Collections.Immutable 1.3.0.
15:04:57 WARNING: Detected package downgrade: System.Collections.Immutable from 1.3.1 to 1.2.1.0 
15:04:57  StackDepthTest (>= 1.0.0) -> CodeAnalysis (>= 1.0.0) -> System.Collections.Immutable (>= 1.3.1) 
15:04:57  StackDepthTest (>= 1.0.0) -> System.Collections.Immutable (>= 1.3.0)
15:04:57 WARNING: Detected package downgrade: System.Collections.Immutable from 1.3.1 to 1.2.1.0 
15:04:57  StackDepthTest (>= 1.0.0) -> CodeAnalysis (>= 1.0.0) -> System.Collections.Immutable (>= 1.3.1) 
15:04:57  StackDepthTest (>= 1.0.0) -> System.Collections.Immutable (>= 1.3.0)
```

**Comment by TyOverby** [comment]
Nevermind, I didn't know there was a newer version of collections immutable in nuget.  Trying that out.

**Comment by TyOverby** [comment]
@jcouv @jaredpar 

The tool isn't being automatically run anywhere, but this PR adds the max-stack-depth-finding tool.

**Pull request merged**